heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,541 @@
+        LOG.severe("Heron job should not finish!");
+
+        break;
+      case TASK_FAILED:
+        LOG.info(String.format("Task with id '%s' FAILED", status.getTaskId().getValue()));
+        handleMesosFailure(taskId);
+
+        break;
+      case TASK_KILLED:
+        LOG.info("Task killed which is supported running. "
+            + "It could be triggered by manual restart command");
+        handleMesosFailure(taskId);
+
+        break;
+      case TASK_LOST:
+        LOG.info(String.format("Task with id '%s' LOST", status.getTaskId().getValue()));
+        handleMesosFailure(taskId);
+
+        break;
+      case TASK_ERROR:
+        LOG.info(String.format("Task with id '%s' ERROR", status.getTaskId().getValue()));
+        handleMesosFailure(taskId);
+
+        break;
+      default:
+        LOG.severe("Unknown TaskState:" + status.getState() + " for taskId: "
+            + status.getTaskId().getValue());
+
+        break;
+    }
+  }
+
+
+  /**
+   * Restart a failed task unless exceeding the retires limitation
+   *
+   * @param taskId the taskId of container to handle
+   */
+  protected void handleMesosFailure(String taskId) {
+    int attempt = TaskUtils.getAttemptForTaskId(taskId);
+    BaseContainer container = containersInfo.get(TaskUtils.getContainerIndexForTaskId(taskId));
+
+    boolean hasAttemptsLeft = attempt < container.retries;
+
+    if (hasAttemptsLeft) {
+      LOG.warning(String.format("Retrying task: %s, attempt: %d", container.name, attempt + 1));
+
+      String newTaskId = TaskUtils.getTaskId(container.name, attempt + 1);
+      scheduleNewTask(newTaskId);
+    } else {
+      LOG.severe("Would not restart the job since it is beyond retries: " + attempt);
+    }
+  }
+
+  /**
+   * Schedule a new task
+   *
+   * @param taskId the taskId of container to handle
+   */
+  protected boolean scheduleNewTask(String taskId) {
+    // Put it to the pending start queue
+    LOG.info(String.format("We are to schedule task: [%s]", taskId));
+
+    // Update the tasksId
+    int containerIndex = TaskUtils.getContainerIndexForTaskId(taskId);
+    tasksId.put(containerIndex, taskId);
+
+    // Re-schedule it
+    toScheduleTasks.add(taskId);
+    LOG.info(String.format("Added task: %s into the to-schedule-tasks queue: ", taskId));
+    return true;
+  }
+
+  @Override
+  public void frameworkMessage(SchedulerDriver schedulerDriver,
+                               Protos.ExecutorID executorID,
+                               Protos.SlaveID slaveID,
+                               byte[] data) {
+    // TODO(mfu): TO handle this message
+    LOG.info("Received framework message.");
+  }
+
+  @Override
+  public void disconnected(SchedulerDriver schedulerDriver) {
+    // TODO(mfu): TO handle this failure
+    LOG.info("Disconnected!");
+  }
+
+  @Override
+  public void slaveLost(SchedulerDriver schedulerDriver, Protos.SlaveID slaveID) {
+    // TODO(mfu): TO handle this failure
```

Comment:
If a slave is lost, that means the tasks on that slave are also lost. Do we need some kind of recovering mechanism? Otherwise, the topology is not working correctly.